pkg/connection/connection.go

```diff
@@ -0,0 +1,420 @@
+	}
+	// No values provided, simple execute is enough
+	if len(args) == 0 {
+		errs.Go(c.executeSimpleWrapper(errctx, query, result))
+	} else {
+		errs.Go(c.executePreparedStatementWrapper(errctx, query, args, result))
+	}
+	err := errs.Wait()
+	close(result)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return <-result, nil
+}
+
+func (c *Connection) executeSimpleWrapper(ctx context.Context, query string, result chan driver.Result) func() error {
+	return func() error {
+		r, err := c.executeSimpleWithResult(ctx, query)
+		if err != nil {
+			return err
+		}
+		result <- r
+		return nil
+	}
+}
+
+func (c *Connection) executePreparedStatementWrapper(ctx context.Context, query string, args []driver.Value, result chan driver.Result) func() error {
+	return func() error {
+		prepResponse := &types.CreatePreparedStatementResponse{}
+		err := c.Send(ctx, &types.CreatePreparedStatementCommand{
+			Command: types.Command{Command: "createPreparedStatement"},
+			SQLText: query,
+		}, prepResponse)
+		if err != nil {
+			return err
+		}
+		resp, err := c.executePreparedStatement(ctx, prepResponse, args)
+		if err != nil {
+			return err
+		}
+		r, err := ToResult(resp)
+		if err != nil {
+			return err
+		}
+		result <- r
+		return nil
+	}
+}
+
+func (c *Connection) getProxy() (*proxy.Proxy, error) {
+	hosts, err := utils.ResolveHosts(c.Config.Host)
+	if err != nil {
+		return nil, err
+	}
+	utils.ShuffleHosts(hosts)
+	return proxy.NewProxy(hosts, c.Config.Port)
+}
+
+func (c *Connection) uploadFiles(ctx context.Context, p *proxy.Proxy, query string) error {
+	paths, err := utils.GetFilePaths(query)
+	if err != nil {
+		return err
+	}
+
+	var files []*os.File
+	for _, path := range paths {
+		f, ferr := utils.OpenFile(path)
+		if ferr != nil {
+			return ferr
+		}
+		files = append(files, f)
+	}
+
+	err = p.Write(ctx, files, utils.GetRowSeparator(query))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Connection) executeSimpleWithResult(ctx context.Context, query string) (driver.Result, error) {
+	result, err := c.SimpleExec(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	return ToResult(result)
+}
+
+func (c *Connection) SimpleExec(ctx context.Context, query string) (*types.SqlQueriesResponse, error) {
+	command := &types.SqlCommand{
+		Command: types.Command{Command: "execute"},
+		SQLText: query,
+		Attributes: types.Attributes{
+			ResultSetMaxRows: c.Config.ResultSetMaxRows,
+		},
+	}
+	result := &types.SqlQueriesResponse{}
+	err := c.Send(ctx, command, result)
+	if err != nil {
+		return nil, err
+	}
+	if result.NumResults == 0 {
+		return nil, errors.ErrMalformedData
+	}
+	return result, err
+}
+
+func (c *Connection) close(ctx context.Context) error {
+	c.IsClosed = true
+	err := c.Send(ctx, &types.Command{Command: "disconnect"}, nil)
+	c.websocket.Close()
+	c.websocket = nil
+	return err
+}
+
+func (c *Connection) Login(ctx context.Context) error {
+	hasCompression := c.Config.Compression
+	c.Config.Compression = false
```

Comment:
Why do we need to deactivate compression here?

Reply:
https://github.com/exasol/websocket-api/blob/master/docs/WebsocketAPIV2.md#compression
The compression is set via login.